internal/eventstream/client.go

```diff
@@ -0,0 +1,421 @@
+// derived from the Client.Ctx and would either stop when this context is done or when the context.CancelFunc is called.
+func (client *Client) startCatchupWorkers() (chan *eventMsg, context.CancelFunc) {
+	startTime := time.Now()
+	eventMsgCh := make(chan *eventMsg)
+
+	// Unfortunately, the errgroup context is hidden, that's why another context is necessary.
+	ctx, cancel := context.WithCancel(client.Ctx)
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	objTypes := []string{"host", "service"}
+	for _, objType := range objTypes {
+		objType := objType // https://go.dev/doc/faq#closures_and_goroutines
+		group.Go(func() error {
+			err := client.checkMissedChanges(groupCtx, objType, eventMsgCh)
+			if err != nil {
+				client.Logger.Errorw("Catch-up-phase event worker failed", zap.String("object type", objType), zap.Error(err))
+			}
+			return err
+		})
+	}
+
+	go func() {
+		err := group.Wait()
+		if err != nil {
+			client.Logger.Errorw("Catching up the API failed", zap.Error(err), zap.Duration("duration", time.Since(startTime)))
+		} else {
+			client.Logger.Infow("Catching up the API has finished", zap.Duration("duration", time.Since(startTime)))
+		}
+
+		cancel()
+		close(eventMsgCh)
+	}()
+
+	return eventMsgCh, cancel
+}
+
+// worker is the Client's main background worker, taking care of event.Event dispatching and mode switching.
+//
+// When the Client is in the catch-up-phase, requested by catchupPhaseRequest, events from the Event Stream API will
+// be cached until the catch-up-phase has finished, while replayed events will be delivered directly.
+//
+// Communication takes place over the eventDispatcherEventStream and catchupPhaseRequest channels.
+func (client *Client) worker() {
+	// eventCache is a subset of event.Event used for caching those in catchupCache below.
+	type eventCache struct {
+		SourceId int64
+		Name     string
+		Type     string
+	}
+
+	var (
+		// catchupEventCh emits events generated during the catch-up-phase from catch-up-workers. It will be closed when
+		// catching up is done, which indicates the select below to switch phases. When this variable is nil, this
+		// Client is in the normal operating phase.
+		catchupEventCh chan *eventMsg
+		// catchupCancel cancels, if not nil, all running catch-up-workers, e.g., when restarting catching-up.
+		catchupCancel context.CancelFunc
+
+		// catchupBuffer holds Event Stream events to be replayed after the catch-up-phase has finished.
+		catchupBuffer = make([]*event.Event, 0)
+		// catchupCache maps eventCache derived from event.Events to API time to skip replaying outdated events.
+		catchupCache = make(map[eventCache]time.Time)
+	)
+
+	// catchupCacheUpdate updates the catchupCache if this eventMsg seems to be the latest of its kind.
+	catchupCacheUpdate := func(ev *eventMsg) {
+		evc := eventCache{ev.event.SourceId, ev.event.Name, ev.event.Type}
+		ts, ok := catchupCache[evc]
+		if !ok || ev.apiTime.After(ts) {
+			catchupCache[evc] = ev.apiTime
+		}
+	}
```

Comment:
`eventCache.Type` is something like state/ack here, not host/service, isn't it? So given that both use the same value for `apiTime`, it shouldn't be necessary to even include this in the cache key.

https://github.com/Icinga/icinga-notifications/blob/a61c6efbebac100a97c44f198c272f2ad6944a87/internal/eventstream/client_api.go#L231-L232

https://github.com/Icinga/icinga-notifications/blob/a61c6efbebac100a97c44f198c272f2ad6944a87/internal/eventstream/client_api.go#L258-L259

Reply:
Yes, `eventCache.Name` is the host or service name (either `$host` for host events or `${host}!${service}` for service events) and `eventCache.Type` the`Event.Type`.
https://github.com/Icinga/icinga-notifications/blob/3b9bef52be6bfba808460da34a56590378c0ddfa/internal/event/event.go#L23
https://github.com/Icinga/icinga-notifications/blob/3b9bef52be6bfba808460da34a56590378c0ddfa/internal/event/event.go#L31-L35

However, when dropping the `.Type` field, a state change would not be distinguishable from an acknowledgement, right?

Reply:
> However, when dropping the `.Type` field, a state change would not be distinguishable from an acknowledgement, right?

Yes, but that's actually a good thing. Currently, there are two scenarios:
1. The checkable was acknowledged, so `catchupCache["$name", "state"]` and `catchupCache["$name", "acknowledgement"]` store the same value, so if it was just `catchupCache["$name"]`, there would be no difference.
2. The checkable was not acknowledged. Now `catchupCache["$name", "acknowledgement"]` won't exist and all acknowledgements from the event stream will be processed, no matter if they have a timestamp before `catchupCache["$name", "state"]` or not. If there's a buffered acknowledgement from the event stream with a timestamp earlier than that, this would mean that this acknowledgment got already dismissed when `/v1/objects` API was queried. So by the time it's processed, it's already obsolete and the processing would happen out of order. If it was just a common value for all event types here, that wouldn't happen.